File: ruby/test/task_master/task_test.rb
require_relative "../test_helper"

class TaskMaster::TaskTest < Minitest::Test
  def setup
    @class_name = TaskMaster::Task
  end

  def test_should_return_id_of_task
    expected = 1
    actual = @class_name.new(
      id: expected.to_s,
      name: "Chore",
      assignee: "John Smith",
      due_on: "2023-10-31",
      status: "In Progress"
    ).id
    assert_equal(expected, actual)
  end

  def test_should_return_name_of_task
    expected = "Chore"
    actual = @class_name.new(
      id: "1",
      name: expected,
      assignee: "John Smith",
      due_on: "2023-10-31",
      status: "In Progress"
    ).name
    assert_equal(expected, actual)
  end

  def test_should_return_assignee_of_task
    expected = "John Smith"
    actual = @class_name.new(
      id: "1",
      name: "Chore",
      assignee: expected,
      due_on: "2023-10-31",
      status: "In Progress"
    ).assignee
    assert_equal(expected, actual)
  end

  def test_should_return_due_on_of_task
    expected = Date.parse("2023-10-31")
    actual = @class_name.new(
      id: "1",
      name: "Chore",
      assignee: "John Smith",
      due_on: expected.to_s,
      status: "In Progress"
    ).due_on
    assert_equal(expected, actual)
  end

  def test_should_return_status_of_task
    expected = "In Progress"
    actual = @class_name.new(
      id: "1",
      name: "Chore",
      assignee: "John Smith",
      due_on: "2023-10-31",
      status: expected
    ).status
    assert_equal(expected, actual)
  end

  def test_should_return_true_if_the_task_has_a_status_that_matches_a_given_list_of_statuses
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "Completed"
      ).status_of?(["Completed"])
    )
  end

  def test_should_return_false_if_the_task_has_a_status_that_differs_a_given_list_of_statuses
    refute(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "Completed"
      ).status_of?(["In Progress"])
    )
  end

  def test_should_return_true_if_the_task_has_a_assignee_that_matches_a_given_list_of_assignees
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "Completed"
      ).assigned_to?(["John Smith"])
    )
  end

  def test_should_return_false_if_the_task_has_a_assignee_that_differs_a_given_list_of_assignees
    refute(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "Completed"
      ).assigned_to?(["Jane Doe"])
    )
  end

  def test_should_return_true_if_the_task_is_complete
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "Completed"
      ).complete?
    )
  end

  def test_should_return_false_if_the_task_is_incomplete
    refute(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "In Progress"
      ).complete?
    )
  end

  def test_should_return_true_if_the_task_is_in_progress
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "In Progress"
      ).in_progress?
    )
  end

  def test_should_return_false_if_the_task_is_not_in_progress
    refute(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "Not Started"
      ).in_progress?
    )
  end

  def test_should_return_true_if_the_task_is_not_started
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "Not Started"
      ).not_started?
    )
  end

  def test_should_return_false_if_the_task_has_been_started
    refute(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: "2023-10-31",
        status: "In Progress"
      ).not_started?
    )
  end

  def test_should_return_true_if_the_task_has_a_due_date_within_the_upcoming_week
    in_next_seven_days = Date.today + 7
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: in_next_seven_days.to_s,
        status: "Completed"
      ).upcoming?
    )
  end

  def test_should_return_false_if_the_task_has_a_due_date_that_is_not_within_the_upcoming_week
    in_next_eight_days = Date.today + 8
    refute(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: in_next_eight_days.to_s,
        status: "Completed"
      ).upcoming?
    )
  end

  def test_should_return_true_if_the_task_has_a_due_date_that_is_past_due
    yesterday = Date.today - 1
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: yesterday.to_s,
        status: "In Progress"
      ).overdue?
    )
  end

  def test_should_return_the_number_of_days_until_due_date_as_integer
    expected = 7
    next_week = Date.today + expected
    assert_equal(
      expected,
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: next_week.to_s,
        status: "Completed"
      ).number_of_days_until_due
    )
  end

  def test_should_return_the_number_of_days_until_due_date_should_positive_if_due_on_in_future
    next_week = Date.today + 7
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: next_week.to_s,
        status: "Completed"
      ).number_of_days_until_due.positive?
    )
  end

  def test_should_return_the_number_of_days_until_due_date_should_negative_if_due_on_in_past
    last_week = Date.today - 7
    assert(
      @class_name.new(
        id: "1",
        name: "Chore",
        assignee: "John Smith",
        due_on: last_week.to_s,
        status: "Completed"
      ).number_of_days_until_due.negative?
    )
  end
end

File: javascript/test/Task.test.js
const path = require('path');
const Task = require(path.join(__dirname, '../lib/Task'));

beforeEach(() => {
	className = Task;
});

test("should return id of task", () => {
	const expected = 1;
	const actual = new className(
		{ id: String(expected),
		name: "Chore",
		assignee: "John Smith",
		dueOn: "2023-10-31",
		status: "In Progress" }
	).id
	expect(actual).toStrictEqual(expected);
});

test("should return name of task", () => {
	const expected = "Chore";
	const actual = new className(
		{ id: "1",
		name: expected,
		assignee: "John Smith",
		dueOn: "2023-10-31",
		status: "In Progress" }
	).name
	expect(actual).toStrictEqual(expected);
});

test("should return assignee of task", () => {
	const expected = "John Smith";
	const actual = new className(
		{ id: "1",
		name: "Chore",
		assignee: expected,
		dueOn: "2023-10-31",
		status: "In Progress" }
	).assignee
	expect(actual).toStrictEqual(expected);
});

test("should return due on of task", () => {
	const expected = new Date(Date.parse("2023-10-31"));
	const actual = new className(
		{ id: "1",
		name: "Chore",
		assignee: "John Smith",
		dueOn: "2023-10-31",
		status: "In Progress" }
	).dueOn
	expect(actual).toStrictEqual(expected);
});

test("should return status of task", () => {
	const expected = "In Progress";
	const actual = new className(
		{ id: "1",
		name: "Chore",
		assignee: "John Smith",
		dueOn: "2023-10-31",
		status: expected }
	).status
	expect(actual).toStrictEqual(expected);

});

test("should return true if the task has a status that matches a list of given statuses", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Completed" }
		).hasStatusOf(["Completed"])
	).toBeTruthy()
});

test("should return false if the task has a status that matches a list of given statuses", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Completed" }
		).hasStatusOf(["In Progress"])
	).toBeFalsy()

});

test("should return true if the task has a assignee that matches a list of given assignees", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Completed" }
		).isAssignedTo(["John Smith"])
	).toBeTruthy()
});

test("should return false if the task has a assignee that matches a list of given assignees", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Completed" }
		).isAssignedTo(["Jane Doe"])
	).toBeFalsy()
});

test("should return true if the task is complete", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Completed" }
		).isComplete()
	).toBeTruthy()
});

test("should return false if the task is incomplete", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "In Progress" }
		).isComplete()
	).toBeFalsy()
});

test("should return true if the task is in progress", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "In Progress" }
		).isInProgress()
	).toBeTruthy()
});

test("should return false if the task is not in progress", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Not Started" }
		).isInProgress()
	).toBeFalsy()
});

test("should return true if the task is not started", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Not Started" }
		).isNotStarted()
	).toBeTruthy()
});

test("should return false if the task has been started", () => {
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: "2023-10-31",
			status: "Completed" }
		).isNotStarted()
	).toBeFalsy()
});

test("should return true if the task has a due date within the upcoming week", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const today = new Date();
	const inNextSevenDays = new Date(today.getFullYear(), today.getMonth(), today.getDate() + 7)
	const inNextSevenDaysAsString = String(
		inNextSevenDays.getFullYear() + '-' + (inNextSevenDays.getMonth() + 1) + '-' + inNextSevenDays.getDate()
	);
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: inNextSevenDaysAsString,
			status: "Not Started" }
		).isUpcoming()
	).toBeTruthy()
});

test("should return false if the task has a due date not within the upcoming week", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const today = new Date();
  const inNextEightDays = new Date(today.getFullYear(), today.getMonth(), today.getDate() + 8);
	const inNextEightDaysAsString = String(
		inNextEightDays.getFullYear() + '-' + (inNextEightDays.getMonth() + 1) + '-' + inNextEightDays.getDate()
	);
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: inNextEightDaysAsString,
			status: "Completed" }
		).isUpcoming()
	).toBeFalsy()
});

test("should return true if the task has a due date that is past due", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const today = new Date();
  const yesterday = new Date(today.getFullYear(), today.getMonth(), today.getDate() - 1);
	const yesterdayAsString = String(
		yesterday.getFullYear() + '-' + (yesterday.getMonth() + 1) + '-' + yesterday.getDate()
	);
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: yesterdayAsString,
			status: "In Progress" }
		).isOverdue()
	).toBeTruthy()
});

test("should return false if the task has a due date that is not past due", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const today = new Date();
  const tomorrow = new Date(today.getFullYear(), today.getMonth(), today.getDate() + 1);
	const tomorrowAsString = String(
		tomorrow.getFullYear() + '-' + (tomorrow.getMonth() + 1) + '-' + tomorrow.getDate()
	);
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: tomorrowAsString,
			status: "In Progress" }
		).isOverdue()
	).toBeFalsy()
});

test("should return the number of days until due date as integer", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const expected = 7;
	const today = new Date();
  const nextWeek = new Date(today.getFullYear(), today.getMonth(), today.getDate() + expected);
	const nextWeekAsString = String(
		nextWeek.getFullYear() + '-' + (nextWeek.getMonth() + 1) + '-' + nextWeek.getDate()
	);
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: nextWeekAsString,
			status: "Completed" }
		).numberOfDaysUntilDue()
	).toStrictEqual(expected)
});

test("should return the number of days until due date as positive if due on in future", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const today = new Date();
  const nextWeek = new Date(today.getFullYear(), today.getMonth(), today.getDate() + 7);
	const nextWeekAsString = String(
		nextWeek.getFullYear() + '-' + (nextWeek.getMonth() + 1) + '-' + nextWeek.getDate()
	);
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: nextWeekAsString,
			status: "Completed" }
		).numberOfDaysUntilDue()
	).toBeGreaterThan(0)
});

test("should return the number of days until due date as negative if due on in past", () => {
	jest.useFakeTimers().setSystemTime(new Date('2023-11-12T00:00:00Z'));
	const today = new Date();
  const lastWeek = new Date(today.getFullYear(), today.getMonth(), today.getDate() - 7);
	const lastWeekAsString = String(
		lastWeek.getFullYear() + '-' + (lastWeek.getMonth() + 1) + '-' + lastWeek.getDate()
	);
	expect(
		new className(
			{ id: "1",
			name: "Chore",
			assignee: "John Smith",
			dueOn: lastWeekAsString,
			status: "Completed" }
		).numberOfDaysUntilDue()
	).toBeLessThan(0)
});
